src/cmd-init

```diff
@@ -126,13 +181,59 @@ mkdir -p src
             (set +x; cd config && echo -n "Config commit: " && git describe --tags --always --abbrev=42)
             ;;
      esac
-     manifest="config/manifest.yaml"
-     if [[ ! -f "${manifest}" ]]; then
-         echo 1>&2 "Failed to find src/${manifest}"
-         fatal "If using a custom configuration, be sure it has a manifest.yaml."
-     fi
  fi)
 
+# Setup the locally stored init config for a given variant
+setup_variant() {
+    if [[ "${#}" -ne 1 ]]; then
+        echo 1>&2 "Error: This function must be called with the variant name"
+    fi
+    local -r variant="${1}"
+    local -r variant_file="src/config/variant_${variant}"
+
+    if [[ ! -f "${variant_file}" ]]; then
+        echo 1>&2 "Could not find the config for the '${variant} variant (${variant_file})"
+        fatal "If you are using a custom configuration, be sure it has a manifest.yaml & image.yaml."
+    fi
+
+    osver="$(head -1 "${variant_file}")"
```

Comment:
I'm confused by the relationship between "variant" and "osver".  Aren't they basically the same thing?  Why do we write the variant to a file, *and* have what's currently `src/config.json`?

I like the "variant" term, so why not just have e.g. `tmp/variant.json` which contains the argument provided by `cosa init --variant`, and then things dispatch on that?

Another way to say this I guess is...why do we have

```
$ cat variant_scos 
c9s
$
```

in openshift/os today?  Why not just `cosa init --variant=c9s`?

Reply:
We need this level of indirection to be able to change the version in this repo only and not have to update downstream pipelines when we switch RHCOS from one version to the other.

If we only have one level, then we need to specify for each pipeline which version to build and update that synchronously.

With two levels, we can have an `rhcos` variant that resolves to the right os version locally from the repo and does not require a change in the pipelines.

Reply:
It's the same for C9S as we will have the same issue when we move to C10S.

Reply:
I could potentially remove this by using a new set of symlinks instead:
```
manifest-rhcos.yaml -> manifest-rhel-8.6.yaml
image-rhcos.yaml -> image-rhel-8.6.yaml
...
```

But we would still have `rhcos`, `rhel-8.6`, `rhel-9.0`, `scos`, `c9s` as variants.

Reply:
Changing the version would be changing more symlinks.

Reply:
> With two levels, we can have an rhcos variant that resolves to the right os version locally from the repo and does not require a change in the pipelines.


Hmm.  OK so originally you were aiming to also support e.g. "rhel-8.4" and "rhel-8.6" as variants.  For that case, I do see the value in having a level of indirection that has e.g. just "rhel-8" be an alias to whichever is defined in the repository.

But I don't personally think it makes sense to try to have "rhcos" resolve to rhel8 or rhel9 in any kind of transparent way.  We haven't really fleshed this out but honestly I'm quite confident we will need to ship both at the same time.  And *that* is something the pipelines will absolutely need to know about - IMO we should just have a `rhcos8` and `rhcos9` pipeline.  The same for c9s versus c10s etc.

Reply:
BTW, some overlapping debate/discussion in https://github.com/openshift/machine-config-operator/pull/3286#issuecomment-1216374529 :arrow_right:  https://github.com/openshift/machine-config-operator/pull/3294

Reply:
I was thinking more about the 9.0 to 9.1 and 9.2 transition. We likely won't be shipping both 9.0 & 9.2 for a given OCP release but we will still need an easy switch during development.

Reply:
Sure, 9.0 vs 9.2 is the same thing as the "rhel-8.4" versus "rhel-8.6" case.  I do see the value in a `rhel9` abstraction over that for the same reason.

I think concretely what I'd propose is dropping the indirection by default - variant names are the same thing as "osver"s.  And I'd propose e.g. "rhel8"/"rhcos8" and "rhel9"/"rhcos9" for those names.  In order to support being more specific when we need to, we can have "rhel8"/"rhcos8" be a *symbolic link* to "rhel84"/"rhcos84".

Reply:
Suggestion: https://github.com/travier/os/tree/cosa-init-variant-test-wip (see last commit https://github.com/travier/os/commit/a19bdc7f8e78db47dcf9fc0d97dff13dbe776f48) to drop the indirection but keep the benefit of having both implicit and explicit variants.

Reply:
I'll do a `s/rhcos8/rhcos-8` for consistency.

Reply:
Yep thanks, looks cleaner to me!

I have no strong opinion on `rhcos8` versus `rhcos-8`.

Reply:
It's worth noting though that the current name in the OCP release payload is `rhel-coreos-8` - we could also match that for consistency.

Reply:
Great. I'll update this one then. We can bikeshed on the name in the os PR 🙂.